rewrite-java/src/main/java/org/openrewrite/java/search/SemanticallyEqual.java

```diff
@@ -0,0 +1,110 @@
+package org.openrewrite.java.search;
+
+import org.openrewrite.Incubating;
+import org.openrewrite.Tree;
+import org.openrewrite.java.AbstractJavaSourceVisitor;
+import org.openrewrite.java.tree.J;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+@Incubating(since = "5.4.4")
+public class SemanticallyEqual extends AbstractJavaSourceVisitor<Boolean> {
+    private final Tree tree;
+
+    public SemanticallyEqual(Tree tree) {
+        this.tree = tree;
+    }
+
+    @Override
+    public Boolean defaultTo(Tree t) {
+        return true;
+    }
+
+    @Override
+    public Boolean reduce(Boolean r1, Boolean r2) {
+        return r1 && r2;
+    }
+
+    @Override
+    public Boolean visitAnnotation(J.Annotation otherAnnotation) {
+        if(tree instanceof J.Annotation) {
+            J.Annotation annotation = (J.Annotation) tree;
+            AtomicInteger index = new AtomicInteger(0);
+
+            return new SemanticallyEqual(annotation.getAnnotationType()).visit(otherAnnotation.getAnnotationType()) &&
+                otherAnnotation.getArgs() != null ?
+                annotation.getArgs().getArgs().stream().allMatch(arg ->
+                    (otherAnnotation.getArgs().getArgs() != null &&
+                        otherAnnotation.getArgs().getArgs().size() > index.get()) ?
+                        new SemanticallyEqual(arg)
+                                .visit(otherAnnotation.getArgs().getArgs().get(index.getAndIncrement())) :
+                        false
+                ) :
+                annotation.getArgs() == null;
+        }
+
+        return false;
+    }
+
+    @Override
+    public Boolean visitIdentifier(J.Ident otherIdent) {
+        if(tree instanceof J.Ident) {
+            J.Ident ident = (J.Ident) tree;
+
+            return ident.getSimpleName().equals(otherIdent.getSimpleName()) &&
+                (ident.getType() != null ?
+                    ident.getType().equals(otherIdent.getType()) :
+                    otherIdent.getType() == null);
+        }
+
+        return false;
+    }
+
+    @Override
+    public Boolean visitFieldAccess(J.FieldAccess otherFieldAccess) {
+        if(tree instanceof J.FieldAccess) {
+            J.FieldAccess fieldAccess = (J.FieldAccess) tree;
+
+            // if the field access is a class literal
+            if(fieldAccess.getSimpleName().equals("class")) {
+                if(!otherFieldAccess.getSimpleName().equals("class")) {
+                    return false;
+                }
+                else {
+                    return fieldAccess.getTarget().getType() == null ||
+                            fieldAccess.getTarget().getType().equals(otherFieldAccess.getTarget().getType());
```

Comment:
If the `fieldAccess.getTarget().getType()` is null, but `otherFieldAccess.getTarget().getType()` is non-null this would return `true`, right? Double check that this condition is handled correctly. I see a few other instances of this pattern within this file, double check those as well.